packages/vike-vue/src/renderer/onRenderClient.ts

```diff
@@ -14,6 +14,16 @@ const onRenderClient: OnRenderClientAsync = async (pageContext): ReturnType<OnRe
 
     const container = document.getElementById('app')!
     const ssr = container.innerHTML !== ''
+
+    if (!ssr) {
+      // insert injected HTML at start / end of body
+      const bodyHtmlStart = pageContext.config.bodyHtmlStart?.(pageContext)
+      if (bodyHtmlStart) document.body.insertAdjacentHTML('afterbegin', bodyHtmlStart)
+
+      const bodyHtmlEnd = pageContext.config.bodyHtmlEnd?.(pageContext) ?? '<div id="teleported"></div>'
```

Comment:
I would go with a default empty `bodyHtmlEnd` - I suppose some don't use teleport, and for the ones that use it the probability of teleporting to a `#teleported` target is small :)
So people would need to define a custom `bodyHtmlEnd` in config most of the time.

Reply:
If fact I don't think this should be here at all - the template is already created in another place.

Maybe i misunderstand how it works.

Reply:
When not using SSR we don't get the output from `onRenderHTML` so we do need to do this in here as well.

Reply:
But in this case (not using SSR) where is the base html created (who writes and what is the content)?
Because the `bodyHtml{Start,End}` should be added there.
Else we would have to manage a very complicated process of checking if it's already there or not,and keep what is already there or delete what is there and add the new content.

Reply:
@brillout Could you elaborate on this? I was confused as well... Does CSR implicitly depend on a `public/index.html` being defined? Does vike handle this and, if so, where is `<div id="app"></div>` injected in the first place?

Reply:
From https://vike.dev/onRenderClient#spa-vs-ssr the shell html is still rendered in `onRenderHtml` but https://github.com/vikejs/vike-vue/blob/9317a7d30a411da5bd557cc2f377953f351deeb0/packages/vike-vue/src/renderer/onRenderHtml.ts#L25 avoids rendering the content of the app.
So the `bodyHtml{Start,End}` should already be there.

Reply:
I see, thanks!

Reply:
Yes, `bodyHtml{Start,End}` is always there, regardless of SSR. Technically, SSR means that `<div id="app">` is empty. There is always an HTML: the only question is whether it's pre-rendered or not, and that's question is independent of SSR.

> Else we would have to manage a very complicated process of checking if it's already there or not,and keep what is already there or delete what is there and add the new content.

@pdanpdan I think we still have to do this because of config inheritance, see https://github.com/vikejs/vike-vue/pull/88#issuecomment-2135865206. I'm thinking if we manage to remove the previously inserted HTML, that should do the trick.

Reply:
Please validate these assumptions:
- we have `test` page inside both a `public` and `private` ancestor (2 tree branches)
- config for `public` has some `bodyHtmlEnd` (`B`) and config for `private` has other `bodyHtmlEnd` (`C`)
- there is also a global config that has another `bodyHtmlEnd` (`A`)
- when using SSR/SSG everything should be fine (full page is created from the beginning) both on initial loading and on navigation from a page to the other 
- when using CSR(SPA) only one initial shell html is created - using which config file (we'll have `A`, `B` or `C`)?
- when using SSR with client side navigation, after navigation from one page to another the `bodyHtmlEnd` content of the initially rendered page will be present